File: sdks/typescript/src/environments.ts
import type { CallOptions, Transport } from "./transport.js";
import type {
  ComputerAction,
  ComputerDisplay,
  ComputerResult,
  CreateRequest,
  EnvironmentInfo,
  Event,
  ExecRequest,
  ExecResult,
  ForkOptions,
} from "./types.js";
import { FuseError } from "./errors.js";
import { requireArg } from "./validate.js";
import { streamEvents } from "./events.js";

/** The server's max page size; list() requests this internally so it walks
 * every page in as few round trips as possible. */
const MAX_PAGE_LIMIT = 200;

/** Filters (and pagination) for environments.list / environments.listPage. */
export interface ListEnvironmentsOptions {
  taskId?: string;
  state?: string;
  hostId?: string;
  /** Page size (server default 50, max 200). Only consulted by listPage —
   * list() always walks every page, so it ignores this and requests the
   * server's max page size. */
  limit?: number;
  /** Opaque cursor from a previous listPage() call's nextCursor. */
  cursor?: string;
}

interface EnvironmentList {
  environments: EnvironmentInfo[];
  next_cursor?: string;
}

/** One page of environments.listPage, plus the cursor to fetch the next one. */
export interface EnvironmentPage {
  environments: EnvironmentInfo[];
  /** Empty/undefined once there are no more results. */
  nextCursor?: string;
}

/** EnvironmentsService manages microVM lifecycle. */
export class EnvironmentsService {
  constructor(private readonly t: Transport) {}

  /** List tracked environments, optionally filtered. Transparently walks
   * every result page. For explicit single-page control (e.g. a cursor from
   * a previous call), use listPage. */
  async list(
    options: ListEnvironmentsOptions = {},
    opts: CallOptions = {},
  ): Promise<EnvironmentInfo[]> {
    const out: EnvironmentInfo[] = [];
    let cursor = options.cursor;
    for (;;) {
      const page = await this.listPage(
        { ...options, limit: MAX_PAGE_LIMIT, cursor },
        opts,
      );
      out.push(...page.environments);
      if (!page.nextCursor) break;
      cursor = page.nextCursor;
    }
    return out;
  }

  /** List one page of tracked environments, optionally filtered. */
  async listPage(
    options: ListEnvironmentsOptions = {},
    opts: CallOptions = {},
  ): Promise<EnvironmentPage> {
    const out = await this.t.json<EnvironmentList>("GET", "/v1/environments", {
      query: {
        task_id: options.taskId,
        state: options.state,
        host_id: options.hostId,
        limit: options.limit ? String(options.limit) : undefined,
        cursor: options.cursor,
      },
      signal: opts.signal,
    });
    return { environments: out.environments ?? [], nextCursor: out.next_cursor };
  }

  /** Fetch a single environment by VM id. */
  async get(vmId: string, opts: CallOptions = {}): Promise<EnvironmentInfo> {
    requireArg(vmId, "vm id");
    return this.t.json<EnvironmentInfo>(
      "GET",
      `/v1/environments/${encodeURIComponent(vmId)}`,
      { signal: opts.signal },
    );
  }

  /** Provision a new environment. Blocks until running or creation fails. */
  async create(body: CreateRequest, opts: CallOptions = {}): Promise<EnvironmentInfo> {
    return this.t.json<EnvironmentInfo>("POST", "/v1/environments", {
      body,
      signal: opts.signal,
    });
  }

  /** Fork an environment into a new one; returns the new environment. */
  async fork(
    vmId: string,
    body: ForkOptions = {},
    opts: CallOptions = {},
  ): Promise<EnvironmentInfo> {
    requireArg(vmId, "vm id");
    return this.t.json<EnvironmentInfo>(
      "POST",
      `/v1/environments/${encodeURIComponent(vmId)}`,
      { query: { action: "fork" }, body, signal: opts.signal },
    );
  }

  /**
   * Run a command inside a running environment's guest and return its exit code
   * with stdout and stderr kept separate.
   *
   * A non-zero exit_code is resolved, not thrown: the command ran and failed.
   * A thrown error means the command could not be run at all.
   *
   * Exec requires the master token.
   *
   * @example
   * const out = await client.environments.exec(id, { cmd: ["ls", "-l"] });
   * if (out.exit_code !== 0) console.error(out.stderr);
   */
  async exec(
    vmId: string,
    body: ExecRequest,
    opts: CallOptions = {},
  ): Promise<ExecResult> {
    requireArg(vmId, "vm id");
    const hasCmd = (body.cmd?.length ?? 0) > 0;
    const hasShell = (body.shell ?? "") !== "";
    if (!hasCmd && !hasShell) {
      throw new FuseError("one of cmd or shell is required");
    }
    if (hasCmd && hasShell) {
      throw new FuseError("cmd and shell are mutually exclusive");
    }
    return this.t.json<ExecResult>(
      "POST",
      `/v1/environments/${encodeURIComponent(vmId)}`,
      { query: { action: "exec" }, body, signal: opts.signal },
    );
  }

  /**
   * Relay one computer-use action to the environment's desktop and return the
   * result, usually carrying a base64 PNG screenshot. Requires an environment
   * booted from a desktop image; on any other image the server answers 503
   * with a reason.
   *
   * @example
   * const res = await client.environments.computer(id, {
   *   action: "left_click",
   *   coordinate: [512, 384],
   * });
   */
  async computer(
    vmId: string,
    action: ComputerAction,
    opts: CallOptions = {},
  ): Promise<ComputerResult> {
    requireArg(vmId, "vm id");
    requireArg(action?.action, "action");
    return this.t.json<ComputerResult>(
      "POST",
      `/v1/environments/${encodeURIComponent(vmId)}/computer`,
      { body: action, signal: opts.signal },
    );
  }

  /** Report whether the environment has a live display and at what geometry.
   * up is false with a reason on an image with no desktop. */
  async computerDisplay(vmId: string, opts: CallOptions = {}): Promise<ComputerDisplay> {
    requireArg(vmId, "vm id");
    return this.t.json<ComputerDisplay>(
      "GET",
      `/v1/environments/${encodeURIComponent(vmId)}/computer`,
      { signal: opts.signal },
    );
  }

  /** Gracefully drain an environment; returns the updated environment. */
  async drain(vmId: string, opts: CallOptions = {}): Promise<EnvironmentInfo> {
    requireArg(vmId, "vm id");
    return this.t.json<EnvironmentInfo>(
      "POST",
      `/v1/environments/${encodeURIComponent(vmId)}`,
      { query: { action: "drain" }, signal: opts.signal },
    );
  }

  /** Rotate the per-VM credentials and push them to the guest. */
  async rotateToken(vmId: string, opts: CallOptions = {}): Promise<void> {
    requireArg(vmId, "vm id");
    await this.t.noContent("POST", `/v1/environments/${encodeURIComponent(vmId)}`, {
      query: { action: "rotate-token" },
      signal: opts.signal,
    });
  }

  /** Forcefully destroy an environment. Idempotent. */
  async destroy(vmId: string, opts: CallOptions = {}): Promise<void> {
    requireArg(vmId, "vm id");
    await this.t.noContent("DELETE", `/v1/environments/${encodeURIComponent(vmId)}`, {
      signal: opts.signal,
    });
  }

  /**
   * Open the SSE event stream for an environment. The returned promise rejects
   * with a FuseApiError on a connect-time failure (e.g. not_found); on success
   * it resolves to an AsyncIterable that yields each Event and ends after a
   * terminal-state event. Pass `opts.signal` to cancel; there is no built-in
   * timeout on the stream.
   *
   * @example
   * for await (const ev of await client.environments.events(id)) {
   *   console.log(ev.state);
   * }
   */
  async events(vmId: string, opts: CallOptions = {}): Promise<AsyncIterable<Event>> {
    requireArg(vmId, "vm id");
    return streamEvents(this.t, vmId, opts.signal);
  }
}

File: sdks/python/src/fuse/_services/environments.py
from __future__ import annotations

from collections.abc import Iterator
from urllib.parse import quote

from .._transport import Transport
from ..types import (
    ComputerAction,
    ComputerDisplay,
    ComputerResult,
    CreateRequest,
    EnvironmentInfo,
    EnvironmentPage,
    Event,
    ExecRequest,
    ExecResult,
    ForkOptions,
)
from .events import stream_events

# the server's max page size; list() requests this internally so it walks
# every page in as few round trips as possible.
_MAX_PAGE_LIMIT = 200


class EnvironmentsService:
    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def list(
        self, *, task_id: str = "", state: str = "", host_id: str = "", cursor: str = ""
    ) -> list[EnvironmentInfo]:
        # returns every environment matching the filters, transparently
        # walking every result page. for explicit single-page control (e.g.
        # a cursor from a previous call), use list_page.
        out: list[EnvironmentInfo] = []
        while True:
            page = self.list_page(
                task_id=task_id,
                state=state,
                host_id=host_id,
                limit=_MAX_PAGE_LIMIT,
                cursor=cursor,
            )
            out.extend(page.environments)
            if not page.next_cursor:
                break
            cursor = page.next_cursor
        return out

    def list_page(
        self,
        *,
        task_id: str = "",
        state: str = "",
        host_id: str = "",
        limit: int = 0,
        cursor: str = "",
    ) -> EnvironmentPage:
        # returns one page of environments matching the filters.
        params: dict[str, str] = {"task_id": task_id, "state": state, "host_id": host_id}
        if limit > 0:
            params["limit"] = str(limit)
        if cursor:
            params["cursor"] = cursor
        resp = self._t.request("GET", "/v1/environments", params=params)
        return EnvironmentPage.model_validate(resp.json())

    def get(self, vm_id: str) -> EnvironmentInfo:
        if not vm_id:
            raise ValueError("vm id is required")
        resp = self._t.request("GET", f"/v1/environments/{quote(vm_id, safe='')}")
        return EnvironmentInfo.model_validate(resp.json())

    def create(self, request: CreateRequest) -> EnvironmentInfo:
        resp = self._t.request("POST", "/v1/environments", body=request)
        return EnvironmentInfo.model_validate(resp.json())

    def drain(self, vm_id: str) -> EnvironmentInfo:
        return self._action(vm_id, "drain")

    def fork(
        self, vm_id: str, options: ForkOptions | None = None
    ) -> EnvironmentInfo:
        if not vm_id:
            raise ValueError("vm id is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}"
        resp = self._t.request(
            "POST", path, params={"action": "fork"}, body=options or ForkOptions()
        )
        return EnvironmentInfo.model_validate(resp.json())

    def computer(self, vm_id: str, action: ComputerAction) -> ComputerResult:
        # relays one computer-use action to the environment's desktop and
        # returns the result, usually carrying a base64 png screenshot.
        # requires an environment booted from a desktop image; on any other
        # image the server answers 503 with a reason.
        if not vm_id:
            raise ValueError("vm id is required")
        if not action.action:
            raise ValueError("action is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}/computer"
        resp = self._t.request("POST", path, body=action)
        return ComputerResult.model_validate(resp.json())

    def computer_display(self, vm_id: str) -> ComputerDisplay:
        # reports whether the environment has a live display and at what
        # geometry. up is false with a reason on an image with no desktop.
        if not vm_id:
            raise ValueError("vm id is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}/computer"
        resp = self._t.request("GET", path)
        return ComputerDisplay.model_validate(resp.json())

    def exec(self, vm_id: str, request: ExecRequest) -> ExecResult:
        # runs a command inside a running environment's guest. requires the
        # master token.
        #
        # a non-zero exit_code is returned, not raised: the command ran and
        # failed. an ApiError means the command could not be run at all.
        if not vm_id:
            raise ValueError("vm id is required")
        has_cmd = bool(request.cmd)
        has_shell = bool(request.shell)
        if not has_cmd and not has_shell:
            raise ValueError("one of cmd or shell is required")
        if has_cmd and has_shell:
            raise ValueError("cmd and shell are mutually exclusive")
        path = f"/v1/environments/{quote(vm_id, safe='')}"
        resp = self._t.request("POST", path, params={"action": "exec"}, body=request)
        return ExecResult.model_validate(resp.json())

    def rotate_token(self, vm_id: str) -> None:
        if not vm_id:
            raise ValueError("vm id is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}"
        self._t.request("POST", path, params={"action": "rotate-token"})

    def destroy(self, vm_id: str) -> None:
        if not vm_id:
            raise ValueError("vm id is required")
        self._t.request("DELETE", f"/v1/environments/{quote(vm_id, safe='')}")

    def events(self, vm_id: str) -> Iterator[Event]:
        # opens the sse stream and yields Event values. the iterator ends
        # cleanly on eof, after a terminal-state event, or after a final
        # Event whose err is set on a stream-level failure.
        return stream_events(self._t, vm_id)

    def _action(self, vm_id: str, action: str) -> EnvironmentInfo:
        if not vm_id:
            raise ValueError("vm id is required")
        if not action:
            raise ValueError("action is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}"
        resp = self._t.request("POST", path, params={"action": action})
        return EnvironmentInfo.model_validate(resp.json())
